File: usvm-ts/src/test/resources/samples/operators/Neg.ts
// @ts-nocheck
// noinspection JSUnusedGlobalSymbols

class Neg {
    negateNumber(a: number): number {
        let res = -a;
        if (Number.isNaN(a)) return res; // -NaN == NaN
        if (a == 0) return res; // -0 == 0
        if (a > 0) return res; // -(x>0) == (x<0)
        if (a < 0) return res; // -(x<0) == (x>0)
        // unreachable
    }

    negateBoolean(a: boolean): number {
        let res = -a;
        if (a) return res; // -true == -1
        if (!a) return res; // -false == -0
        // unreachable
    }

    negateUndefined(): number {
        let x = undefined;
        return -x; // -undefined == NaN
    }

    negateNull(): number {
        let x = null;
        return -x; // -null == -0
    }

    negateString(a: string): number {
        let res = -a;

        if (a === "") return res; // -"" == -0
        if (a === "0") return res; // -"0" == -0
        if (a === "1") return res; // -"1" == -1
        if (a === "-42") return res; // -"-42" == 42
        if (a === "NaN") return res; // -"NaN" == NaN
        if (a === "Infinity") return res; // -"Infinity" == -Infinity
        if (a === "-Infinity") return res; // -"Infinity" == Infinity

        if (a === "hello") return res; // -"hello" == NaN
        if (a === "true") return res; // -"true" == NaN
        if (a === "false") return res; // -"false" == NaN
        if (a === "undefined") return res; // -"undefined" == NaN
        if (a === "null") return res; // -"null" == NaN
        if (a === "(42)") return res; // -"(42)" == NaN

        if (a === "{}") return res; // -"{}" == NaN
        if (a === "{foo: 42}") return res; // -"{foo: 42}" == NaN
        if (a === "[]") return res; // -"[]" == NaN
        if (a === "[42]") return res; // -"[42]" == 42

        return res; // can be either NaN or a number
    }
}

File: usvm-ts/src/test/kotlin/org/usvm/samples/operators/Neg.kt
package org.usvm.samples.operators

import org.jacodb.ets.model.EtsScene
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import org.usvm.api.TsTestValue
import org.usvm.util.TsMethodTestRunner
import org.usvm.util.eq
import org.usvm.util.isNaN

class Neg : TsMethodTestRunner() {
    private val tsPath = "/samples/operators/Neg.ts"

    override val scene: EtsScene = loadScene(tsPath)

    @Test
    fun `test negateNumber`() {
        val method = getMethod("negateNumber")
        discoverProperties<TsTestValue.TsNumber, TsTestValue.TsNumber>(
            method = method,
            { a, r ->
                // -NaN = NaN
                a.isNaN() && r.isNaN()
            },
            { a, r ->
                // -0 = -0
                (a eq 0) && (r eq -0.0)
            },
            { a, r ->
                // -positive number = negative number
                (a.number > 0) && (r.number < 0)
            },
            { a, r ->
                // -negative number = positive number
                (a.number < 0) && (r.number > 0)
            },
            invariants = arrayOf(
                { a, r ->
                    if (a.isNaN()) {
                        r.isNaN()
                    } else {
                        r.number eq -a.number
                    }
                },
            )
        )
    }

    @Test
    fun `test negateBoolean`() {
        val method = getMethod("negateBoolean")
        discoverProperties<TsTestValue.TsBoolean, TsTestValue.TsNumber>(
            method = method,
            { a, r ->
                // -true = -1
                a.value && (r eq -1)
            },
            { a, r ->
                // -false = -0
                !a.value && (r eq 0)
            },
        )
    }

    @Test
    fun `test negateUndefined`() {
        val method = getMethod("negateUndefined")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r ->
                // -undefined = NaN
                r.isNaN()
            },
        )
    }

    @Test
    fun `test negateNull`() {
        val method = getMethod("negateNull")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r ->
                // -null = -0
                r eq 0
            },
        )
    }

    @Disabled("Input strings are not supported yet")
    @Test
    fun `test negateString`() {
        val method = getMethod("negateString")
        discoverProperties<TsTestValue.TsString, TsTestValue.TsNumber>(
            method = method,
            { a, r ->
                // -"" = -0
                a.value == "" && (r eq 0.0)
            },
            { a, r ->
                // -"0" = -0
                a.value == "0" && (r eq 0.0)
            },
            { a, r ->
                // -"1" = -1
                a.value == "1" && (r eq -1.0)
            },
            { a, r ->
                // -"-42" = 42
                a.value == "-42" && (r eq 42.0)
            },
            { a, r ->
                // -"NaN" = NaN
                a.value == "NaN" && r.isNaN()
            },
            { a, r ->
                // -"Infinity" = -Infinity
                a.value == "Infinity" && (r eq Double.NEGATIVE_INFINITY)
            },
            { a, r ->
                // -"-Infinity" = Infinity
                a.value == "-Infinity" && (r eq Double.POSITIVE_INFINITY)
            },
            { a, r ->
                // -"hello" = NaN
                a.value == "hello" && r.isNaN()
            },
            { a, r ->
                // -"true" = NaN
                a.value == "true" && r.isNaN()
            },
            { a, r ->
                // -"false" = NaN
                a.value == "false" && r.isNaN()
            },
            { a, r ->
                // -"undefined" = NaN
                a.value == "undefined" && r.isNaN()
            },
            { a, r ->
                // -"null" = NaN
                a.value == "null" && r.isNaN()
            },
            { a, r ->
                // -"(42)" = NaN
                a.value == "(42)" && r.isNaN()
            },
            { a, r ->
                // -"{}" = NaN
                a.value == "{}" && r.isNaN()
            },
            { a, r ->
                // -"{foo: 42}" = NaN
                a.value == "{foo: 42}" && r.isNaN()
            },
            { a, r ->
                // -"[]" = NaN
                a.value == "[]" && r.isNaN()
            },
            { a, r ->
                // -"[42]" = 42
                a.value == "[42]" && (r eq 42.0)
            },
            invariants = arrayOf(
                { _, _ -> true }
            )
        )
    }
}
